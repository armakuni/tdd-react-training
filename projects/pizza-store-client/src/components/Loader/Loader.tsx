import { ReactElement } from 'react';
import useLoader from '../../hooks/useLoader';

export type LoadFunction<ReturnType> = () => Promise<ReturnType>

interface LoaderProps<ReturnType> {
  loader: LoadFunction<ReturnType>;
  children: (data: ReturnType) => ReactElement;
}

export default function Loader<ReturnType>({ loader, children }: LoaderProps<ReturnType>): ReactElement {
  const result = useLoader(loader);

  switch (result.state) {
    case 'loaded':
      return children(result.data);
    case 'error':
      return <div>{result.error}</div>;
    case 'loading':
    default:
      return <div>Loading</div>;
  }
}
