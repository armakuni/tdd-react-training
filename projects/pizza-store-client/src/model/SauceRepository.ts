import { Sauce } from './Sauce';

export type FetchSauces = () => Promise<Sauce[]>
