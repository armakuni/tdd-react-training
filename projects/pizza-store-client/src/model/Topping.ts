export type ToppingID = number;

export interface Topping {
  id: ToppingID;
  name: string;
  price: number;
}
