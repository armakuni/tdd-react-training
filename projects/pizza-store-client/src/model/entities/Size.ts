export type SizeID = string;

export interface Size {
  id: SizeID;
  display: string;
  price: number;
  toppingPriceMultiplier: number;
}
