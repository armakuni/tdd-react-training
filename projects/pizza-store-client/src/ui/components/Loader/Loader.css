.loader {
  border: 6px solid #369;
  /* Light grey */
  border-top: 6px solid #bbb;
  /* Blue */
  border-radius: 50%;
  width: 50px;
  height: 50px;
  animation: spin 2s linear infinite;
  margin-left: auto;
  margin-right: auto;
  margin-top: 20px;
  margin-bottom: 20px;
}

@keyframes spin {
  0% {
    transform: rotate(0deg);
  }

  100% {
    transform: rotate(360deg);
  }
}

.loader-error {
  color: #c33;
  text-align: center;
}

.loader-error__title {
  color: #c33;
  font-weight: bold;
}
