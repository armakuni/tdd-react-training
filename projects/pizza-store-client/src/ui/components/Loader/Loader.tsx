import { ReactElement } from 'react';
import useLoader from '../../hooks/useLoader';
import './Loader.css';

export type LoadFunction<ReturnType> = () => Promise<ReturnType>

interface LoaderProps<ReturnType> {
  loader: LoadFunction<ReturnType>;
  children: (data: ReturnType) => ReactElement;
}

export default function Loader<ReturnType>({ loader, children }: LoaderProps<ReturnType>): ReactElement {
  const result = useLoader(loader);

  switch (result.state) {
    case 'loaded':
      return children(result.data);
    case 'error':
      return (
        <div className="loader-error">
          <div className="loader-error__title">An Error Occurred:</div>
          <div className="loader-error__message">{result.error}</div>
        </div>
      );
    case 'loading':
    default:
      return <div className="loader" data-testid="loader" />;
  }
}
