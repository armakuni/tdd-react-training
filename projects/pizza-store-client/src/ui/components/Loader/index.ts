import SizeLoader from './Loader';

export default SizeLoader;
