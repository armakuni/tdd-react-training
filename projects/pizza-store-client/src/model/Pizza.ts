import { Sauce } from './Sauce';
import { Size } from './Size';
import { ToppingID } from './Topping';

export interface Pizza {
  readonly size: Size | undefined;
  readonly sauce: Sauce | undefined;
  readonly toppings: Set<ToppingID>;
}

export function create(): Pizza {
  return {
    size: undefined,
    sauce: undefined,
    toppings: new Set(),
  };
}

export function setSize(pizza: Pizza, size: Size): Pizza {
  return { ...pizza, size };
}

export function setSauce(pizza: Pizza, sauce: Sauce): Pizza {
  return { ...pizza, sauce };
}

export function setToppings(pizza: Pizza, toppings: Set<ToppingID>): Pizza {
  return { ...pizza, toppings };
}
