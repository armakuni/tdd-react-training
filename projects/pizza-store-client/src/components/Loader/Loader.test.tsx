import { ReactElement } from 'react';
import {
  render, RenderResult, screen,
} from '@testing-library/react';
import Loader, { LoadFunction } from './Loader';
import { Size } from '../../model/Size';

function renderSizeLoader(
  loader: LoadFunction<string[]>,
  children: (results: string[]) => ReactElement,
) {
  return render(
    <Loader loader={loader}>
      {children}
    </Loader>,
  );
}

function mockChildren(sizes: Size[]): ReactElement {
  return <>{sizes.map((size) => <span key={size}>{size}</span>)}</>;
}

describe('Loader', () => {
  let view: RenderResult;

  describe('when loaded', () => {
    beforeEach(async () => {
      const loader: LoadFunction<string[]> = () => new Promise((resolve) => {
        resolve(['big', 'small']);
      });
      view = renderSizeLoader(loader, mockChildren);
      await screen.findByText('big');
    });

    it('renders the children with the data', () => {
      expect(screen.getByText('big')).toBeVisible();
      expect(screen.getByText('small')).toBeVisible();
      expect(screen.queryByText('Loading')).not.toBeInTheDocument();
    });

    it('matches the snapshot', () => {
      expect(view).toMatchSnapshot();
    });
  });

  describe('when an error occurred', () => {
    beforeEach(async () => {
      const loader: LoadFunction<string[]> = () => new Promise((_resolve, reject) => {
        reject(new Error('there was an error'));
      });
      view = renderSizeLoader(loader, mockChildren);
      await screen.findByText('there was an error');
    });

    it('renders an error message', () => {
      expect(screen.getByText('there was an error')).toBeVisible();
      expect(screen.queryByText('big')).not.toBeInTheDocument();
      expect(screen.queryByText('small')).not.toBeInTheDocument();
      expect(screen.queryByText('Loading')).not.toBeInTheDocument();
    });

    it('matches the snapshot', () => {
      expect(view).toMatchSnapshot();
    });
  });

  describe('when loading', () => {
    beforeEach(async () => {
      const loader: LoadFunction<string[]> = () => new Promise(() => {
        /* never resolve */
      });
      view = renderSizeLoader(loader, mockChildren);
      await screen.findByTestId('loader');
    });

    it('renders the loading message', () => {
      expect(screen.getByTestId('loader')).toBeVisible();
      expect(screen.queryByText('big')).not.toBeInTheDocument();
      expect(screen.queryByText('small')).not.toBeInTheDocument();
    });

    it('matches the snapshot', () => {
      expect(view).toMatchSnapshot();
    });
  });
});
