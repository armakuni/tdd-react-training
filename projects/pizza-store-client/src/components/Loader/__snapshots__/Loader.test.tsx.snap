// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader when an error occurred matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        there was an error
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      there was an error
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Loader when loaded matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span>
        big
      </span>
      <span>
        small
      </span>
    </div>
  </body>,
  "container": <div>
    <span>
      big
    </span>
    <span>
      small
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Loader when loading matches the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        Loading
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      Loading
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
